### DISCUSSION TITLE
github pages: how to mix quarto website with some html pages not to be rendered in quarto?

### DISCUSSION THREAD

**Discussion by aborruso** [quarto-dev/quarto-cli#11725] [Q&A] [answered]
### Description

Hi,
I would like to build a quarto website, and render it via github pages.

I'm starting to build it, and github use `docs` folder to read it `output-dir: docs`.

In this folder, I need to have some subfolders to store HTML files that I want to publish as they are on GitHub, without having Quarto handle the rendering.
I would like to call this folder `garage` and I will put it in `docs`. In this folder "docs/garage" I will put various subfolders, and in these I will have HTML files. 
How do I make sure that I do not have them deleted every time I render with `quarto render`? Do I have to add in the project configuration this glob in project?

```
  resources:
    -  "garage/*"
```

**Comment by mcanouil** [comment]
You can indeed resources but that not a requirement depending on how you render.

**Reply by aborruso** [reply]
Hi @mcanouil and thank you. 

I usually run a default rendering. And I have set `docs` as `output-dir`.
How do I disable deletion of a specific folder located in docs?

**Reply by mcanouil** [reply]
I don't believe `quarto render` in a project erase the content (I am doing that for a client where R packages documentation are committed in `docs` alongside the Quarto documentation.

And using `resources` can ensure it does not happen.

**Reply by aborruso** [reply, marked as answer]
> I don't believe `quarto render` in a project erase the content (I am doing that for a client where R packages documentation are committed in `docs` alongside the Quarto documentation.

Using quarto 1.5, it deletes all I have in docs folder.
I solved it using the mode I had thought of, which is using resources. If you can think of a more convenient way, if you have any alternative advice, I always read you with great interest.

Thank you

**Reply by mcanouil** [reply]
Using `resources` is the safest and more robust way to keep contents not generated by Quarto.

**Reply by jimjam-slam** [reply]
An alternate way you could do this is to keep the HTML files outside `docs` (eg. `garage/some_file.html` instead of `docs/garage/some_file.html`), and have a post-render script copy them into `docs`.

That said, my understanding is that adding `resources: ["garage/*"]` does something similar: it copies files from `garage` to `docs/garage` when you render. If you're storing the source HTML files in `docs/garage` and trying to protect them with `resources`, will that work?

**Reply by cderv** [reply]
> I don't believe quarto render in a project erase the content

`quarto render` does delete `output-dir` content before rendering for project
https://github.com/quarto-dev/quarto-cli/blob/3f492cc7df25df96d3a83456f72fab704b05d644/src/command/render/project.ts#L303-L323

> I solved it using the mode I had thought of, which is using resources.

This is like @jimjam-slam described. Using `resources` config is the way to tell quarto to copy listed content into the output directory. This is the way to tell Quarto how to include some resource in the resulting website (about [site resources](https://quarto.org/docs/websites/website-tools.html#site-resources)). So this seems to me how to do it. You could also do it as post-render script if you prefer to control the inclusion of those content in a specific way.

**Reply by mcanouil** [reply]
So `--clean` is set by default thus `--no-clean` is required if one wants to disable this feature.

**Reply by cderv** [reply]
> So --clean is set by default thus --no-clean is required if one wants to disable this feature.

`book`, `manuscript` and `website` projects do cleaning by default yes, and `flags.clean = true` is indeed the implicit default. So yes using `--no-clean` will not remove the output directory. Though, it could leave some stale pages of resources.

**Reply by mcanouil** [reply]
> If you're storing the source HTML files in docs/garage and trying to protect them with resources, will that work?

Actually, it does not.

```sh
quarto create project website demo demo --no-open
cd demo
sed -i '' '3i\
  resources: ["something"]
' _quarto.yml
quarto render
quarto render --output-dir _site/something
tree -L1 _site
quarto render
tree -L1 _site
```

So, if the content is not in the root directory, it will be deleted by Quarto.

**Reply by mcanouil** [reply]
To overcome this in a side project, I pull the "external" directory (i.e., a directory named `preview`) back into the site directory before publishing.
I did not use `--no-clean` because I still want Quarto to cleanup outdated files.

https://github.com/mcanouil/quarto-pr-preview

**Reply by aborruso** [reply]
> That said, my understanding is that adding `resources: ["garage/*"]` does something similar: it copies files from `garage` to `docs/garage` when you render. If you're storing the source HTML files in `docs/garage` and trying to protect them with `resources`, will that work?

The way to do it (or rather, the way I'm testing and it works) is in example to have this kind of config:

```yml
project:
  type: website
  output-dir: docs
  resources:
    -  "garage/*/"
```

`garage` is a folder in the root of my project and in it I put the html files that I want them not to be managed by quarto.
When I run the quarto render, it copies the `garage` folder to the `docs` folder.

**Reply by mcanouil** [reply]
Indeed, if your files are anywhere but in the `output-dir` at the time of `quarto render`, then the files will be copied by Quarto.
